### ISSUE TITLE
Enable Perfect Forward Secrecy

### ISSUE THREAD

**Issue by tlvenn** [containous/traefik#811]
It would be nice if Traefik would enable Perfect Forward Secrecy for https entryPoints.

**Comment by errm** [comment]
I think is just a case of configuring the cipher suites correctly.  This functionality was added in https://github.com/containous/traefik/pull/690/files.

If my understanding is correct, choosing just the ECDHE based ciphers should be enough.

**Comment by errm** [comment]
Thinking about this it might be nice to have some presets for different "recommended" suites. This could be one...

**Comment by yyvess** [comment]
Mozilla propose suites for "Old", "Intermediate" and " Modern" browsers.
https://mozilla.github.io/server-side-tls/ssl-config-generator/

**Comment by trecloux** [comment]
@errm @yyvess The Mozilla recommended configurations do not only cover but also the SSL/TLS version, session settings, ...
Documentation : https://wiki.mozilla.org/Security/Server_Side_TLS#Recommended_configurations
Mozilla provides a JSON file with the recommended configuration, we could use it to generate Traefik predefined/recommended configs
I think that we could introduce new configuration entry
```
[entryPoints.https.tls]
MozillaRecommendedConfiguration = "modern"
```
This would provide default values for `MinVersion`and `CipherSuites`
WDYT ?

**Comment by yyvess** [comment]
Yes good idead, JSON file should downloaded at build time.

**Comment by emilevauge** [comment]
@trecloux, @yyvess I think this is a great idea :)
Maybe using https://github.com/mozilla/tls-observatory/blob/3af7b37832ef7293d5277a068dc0684deacc6789/worker/mozillaEvaluationWorker/mozillaEvaluationWorker.go from https://github.com/mozilla/tls-observatory

**Comment by trecloux** [comment]
@emilevauge Yeah, looks interesting, it's checking latest configuration online and has a backup configuration. 
I can work on this topic and come back with a pull request.

**Comment by emilevauge** [comment]
@trecloux 
>I can work on this topic and come back with a pull request.
:heart_eyes_cat:

**Comment by yyvess** [comment]
@emilevauge @trecloux Personally I recommend to download the JSON from Mozilla at build time inside at runtime.
Download it on runtime can be generate issues if the file format is updated by Mozilla.
Secondly can be a risk if the Mozilla site will be hacked, I have do not see any Md5/Sha to valid the content of this JSON.

**Comment by trecloux** [comment]
Here is a first POC : https://github.com/trecloux/traefik/commit/2bb1e795031c8f18f6c72bd3fabc46c1ca46e9b3
I would be happy to receive some feedback.
Missing : 
* Try to reduce copy/paster from tls-observatory
* Add tests assertions
* Retrieve JSON file at build time
From my point of view, we need also to :
* Support different key types for Let's encrypt
* Add security headers like HSTS (#669)
@yyvess : I think that we could let the user choose if wants to use the configuration retrieved at build time or he wants to use the latest version.

**Comment by emilevauge** [comment]
@trecloux Oh great work :heart_eyes_cat: !
Can I suggest to open a PR with [WIP] in the tile? It will be easier to follow the developments :)

**Comment by yyvess** [comment]
@emilevauge I aggre, can you add an option to disable the remote mode ?

**Comment by ldez** [comment]
Closed by #1276

**Comment by beejei** [comment]
I don't know why @trecloux 's POC has not considered.
There's no other configuration to define the 
- cipher compatibility,
- http/2 activity
consistently.

For instance, #1371 represents the poor usability of configuration.

**Comment by emilevauge** [comment]
@beejei FYI, @trecloux did not have enough time to work on this.
But, you can already configure which cipher you want and the min TLS version in Traefik.

**Comment by rdxmb** [comment]
Another idea is to have put traefik-parameters into bettercrypto, like here they are for nginx: https://bettercrypto.org/#_nginx
